### PR TITLE
Time standardization and slope priors

### PR DESCRIPTION
**Pull request by afmagee42**
This is part PR, part investigation.

## Background

As noted in #47, the priors I implemented in #41 didn't match the way we handle time. I calibrated priors in terms of days, our model was working in terms of standardized time (time with mean 0 and variance 1). This model worked, in that MCMC ran acceptably, but did not work, in that the priors were too informative and the posteriors of the independent divisions model were too strongly influenced.

## Previous work

Simply removing the standardization on time, however, led to less than stellar MCMC mixing of the Independent Divisions model, and pathological behavior of the Hierarchical Divisions model.

Previous experimentation suggested that standardizing time improved MCMC mixing. However, attempting to mix-and-match scales (set priors on real time, infer on standardized, #55) also led to apparently pathologically bad mixing. In hindsight, this is likely because the transformation induces a strong correlation, where we can otherwise parameterize things more orthogonally.

### TL;DR 
We can make MCMC work by using a dense mass matrix instead of a diagonal mass matrix.

I renamed the `HierarchicalDivisonModel` the `HierarchicalCorrelatedDeviationsModel` and implemented a MVN-free, simpler `HierarchicalDivisonModel`. Both of these have "dynamic pooling" as below, and have one small parameter removed from the previous `HierarchicalDivisonModel`.

### What I tried

To debug the hierarchical model, I made a version which could be set to be mathematically equivalent to the Independent model. This required a minor simplification, removing the free parameter `sigma_beta_0` (a scalar on intercept variances allowing some lineages to vary more in intercepts than others), and was facilitated by removing the correlated deviations in slopes (where the independent model was only recovered in a limit as we take a fixed hyperparameter to infinity).

This enabled me to rewrite the model as described below. Using that, I was able to get a version of the Hierarchical model which could have parameters fixed to recapitulate the Independent Divisions model. I found that when I re-wrote this model to use plates (plating first over lineage, then division) I could obtain MCMC which mixed as well as the Independent Divisions model. Which is to say, not _well_, but not so poorly that we couldn't just run it longer and obtain satisfactory convergence diagnostics.

As I increased the strength of the pooling (either letting the model infer it, where it is very strong, or fixing it to 0.5, which is what we implicitly had before), mixing got worse. I tried using `numpyro.infer.reparam.LocScaleReparam` but this did not appear to have any effect. I imagine that is because the model was already written in the space that parameterization seeks (location 0, common scale 1).

I then tried altering the NUTS settings to use a dense mass matrix, which can account for posterior correlations in parameters. With this turned on, both the Independent and stripped-down Hierarchical models work much better, if slower. (I think the efficiency is worth the tradeoff, such that to get equivalent ESS we'd need to spend more time running the model. But I did not check.) 

Lastly, I tried adding back in the correlated deviations of the slopes. With a dense mass matrix, this model can also be sampled.

I did not try adding back `sigma_beta_0` as while it is perfectly sensible, the uncorrelated-deviations, dynamic pooling hierarchical model seems to be the simplest starting point for pooled model development and we might as well test what works and doesn't in comparison.

### What I think I learned

Partial pooling makes the posterior geometry gnarlier, plates can be useful where applicable, noncentered parameterizations are good, and a good (bad) prior can be bad (good) for mixing.

### Dynamic pooling
All slopes and intercepts here are taken to be on the logit scale and absent a reference category, as we have been doing.

For a model with uncorrelated slope deviations, for location $g$ and lineage $\ell$, both intercepts $\boldsymbol{\beta}\_0$ and slopes $\boldsymbol{\beta}\_1$ can be written as
```math
\beta_{g \ell} := \mu_{\ell} +Z_{g \ell}
```
with
```math
\mu_{\ell} \sim \text{Normal}(0, \sigma_{\mu}^2)
```
and
```math
Z_{g \ell} \sim \text{Normal}(0, \sigma_Z^2)
```
and fixed $\sigma$.

The marginal prior variance is $\text{V}[\beta_{g \ell}] = \text{V}[\mu_{\ell}]  + \text{V}[Z_{g \ell}] = \sigma_Z + \sigma_{\mu}$. 

As #41 shows, it is really $\text{V}[\beta_{g \ell}]$ that we have any intuition about.

We can recover the Independent Divisions model by fixing $\mu_{\ell}$ to 0, such that all of the variance in $\beta_{g \ell}$ is due to $Z_{g \ell}$. We can also imagine on the other side a fully pooled model, in which  all of the variance in $\beta_{g \ell}$ is due to $\mu_{\ell}$. The "strength" of the pooling then can be described by how much of the variance comes from either component, which suggests that we can in fact re-write all three models (unpooled, fully pooled, partially pooled) as
```math
\beta_{g \ell} := \mu_{\ell} + Z_{g \ell}
```
with
```math
\mu_{\ell} \sim \text{Normal}(0, \rho \sigma_{\beta}^2)
```
and
```math
Z_{g \ell} \sim \text{Normal}(0, (1 - \rho) \sigma_{\beta}^2)
```
and fixed $\sigma_{\beta}$.

We can set $\rho = 1$ to get a fully pooled model, let $\rho$ be inferred, or fix $\rho = 0$ to recover the Independent Divisions model. I have implemented this model with separate pooling for the slopes and intercepts.

We can also apply this framework to the original Hierarchical formulation where differences in slopes from the grand mean are correlated. We have a number of length $\ell$ vectors instead of scalars (bold is not rendering for some reason)
```math
\boldsymbol{\beta}_g = \boldsymbol{\mu} + \boldsymbol{Z}_{g}
```
with (as in the non-correlated model)
```math
\mu_{\ell} \sim \text{Normal}(0, \rho \sigma_{\beta}^2)
```
and
```math
\boldsymbol{Z}_{g} \sim \text{MultivariateNormal}(0, \Sigma)
```
where $\Sigma = \text{diag}((1 - \rho) \sigma_{\beta}, L) \mathbf{R} \text{diag}((1 - \rho) \sigma_{\beta}, L)$, $R$ is a correlation matrix, $\text{diag}((1 - \rho) \sigma_{\beta}, L)$ is an $L x L$ diagonal matrix with all elements equal to $(1 - \rho) \sigma_{\beta}$, and
```math
R \sim \text{LKJ}(\eta)
```
As we were already working with our covariance matrix decomposed into a diagonal matrix of standard deviations and a correlation matrix, the $\rho, (1 - \rho)$ formulation is plug and play here too.

### CODE REVIEW

**Comment by afmagee42** [comment]
I'm running all three time points in `retrospective_forecasting` with all the models including both hierarchical ones. So far both Early and Late 21L have run, and in both the non-correlated model produces _slightly_ (like, on the order of a percent or so) better evaluation scores for all 4 metrics.

**Comment by afmagee42** [comment]
And on Mid 21L, the tally is 2 where the correlated hierarchical model is substantially better, 2 where it scores NaN (the proportions are NaN, the counts aren't), and the uncorrelated hierarchical loses to the Independent on both count scores. So, nothing's simple.

**Comment by afmagee42** [comment]
I had missed the fact that in Mid 21L, the non-correlated model also failed to mix, with 1 chain very stuck and 1 chain slow-mixing.
